crates/starknet-types-core/src/curve/affine_point.rs

```diff
@@ -0,0 +1,45 @@
+use crate::curve::curve_errors::CurveError;
+use crate::felt::Felt;
+use lambdaworks_math::cyclic_group::IsGroup;
+use lambdaworks_math::elliptic_curve::short_weierstrass::curves::stark_curve::StarkCurve;
+use lambdaworks_math::elliptic_curve::short_weierstrass::point::ShortWeierstrassProjectivePoint;
+use lambdaworks_math::elliptic_curve::traits::FromAffine;
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct AffinePoint(pub(crate) ShortWeierstrassProjectivePoint<StarkCurve>);
+
+impl AffinePoint {
+    pub fn new(x: Felt, y: Felt) -> Result<AffinePoint, CurveError> {
+        Ok(Self(ShortWeierstrassProjectivePoint::from_affine(
+            x.0, y.0,
+        )?))
+    }
+
+    /// The point at infinity.
+    pub fn identity() -> AffinePoint {
+        Self(ShortWeierstrassProjectivePoint::neutral_element())
+    }
+
+    /// Returns the `x` coordinate of the point.
+    pub fn x(&self) -> Felt {
+        Felt(*self.0.x())
+    }
+
+    /// Returns the `y` coordinate of the point.
+    pub fn y(&self) -> Felt {
+        Felt(*self.0.y())
+    }
```

Comment:
I expect the signature to be `pub fn x(&self) -> &Felt`.
This can be achieved with some raw pointers and a pinch of unsafe code.
Here is a code snippet summarizing how:
```rust
struct W1(u32);

impl W1 {
    fn inner(&self) -> &u32 {
        &self.0
    }
}

struct W2(u32);

struct W3(W1);

impl W3 {
    fn inner(&self) -> &W2 {
     let raw_pointer = self.0.inner() as *const u32 as *const W2;
     unsafe { &*raw_pointer }
    }
}
```

Reply:
Hi guys!
I am not really a fan of adding unsafe code in such core libraries.
Returning the struct instead of the reference is not such a problem, so I think are best choice will be to leave it like that.
What do you think?

Reply:
I personally don't mind unsafe code as long as it's properly commented. However, for something like that, it's not clear whether it would be make a difference or not.

Here's what's to consider according to me:

When returning the type by value, Rust has to allocate 32 bytes on the stack, potentially clobbering some registers while trying to copy the return value into the allocated space. That's generally not good for performance.

When returning a reference, only eight bytes are needed, and they won't probably even be allocated at all because eight bytes fits in a regular CPU register. Moreover, because the function takes a reference in the first place, the function becomes essentially a no-op for the compiler. That's good for performance.

However, simply by marking the function as `#[inline(always)]` or `#[inline]`, the compiler will probably manage to do that optimization anyway,  removing the copy if possible. I know it struggles to do that with huge types (like 256 bytes), but we're well bellow that threshold.

To conclude: unsafe code would make the function fast always. With no unsafe, it's up to the compiler.

Reply:
>  I am not really a fan of adding unsafe code in such core libraries.

On the other hand, performances are utterly important in such core libraries. This method will be called millions of times in any program that uses the lib. We have to do our best the lib consumers.

Reply:
I think the consensus right now (after discusing about this on telegram), is that unsafe code should generally be avoided for this kind of case.

@tdelabro you convined me that it's a good idea to use it though. So after I have some data to back it up (a project with an example where this matters), I'll come back make a pull request.

For now it's perfectly fine to just go with copies, for the small improvement we will gain.